_site/open-source/security/protect-your-sensitive-data-in-git.html

```diff
@@ -4,41 +4,48 @@
ATE KEY-----MIIJKQIBAAKCAgEAoequrqsM42na3OpvBFYOpqvzJumr3/kxJTuluXbPyJzVjMXfd/uhFUJgSqq4AJGOFLLPpQ+9jwfA+WraIxZ9R7p8LgpNdUwKsmGnUvofeD/9Rs1yYZO8EAjl1URLJ379nN+L5KKPS/48Q4iGp57iwuGzrXLHccLyW5+Z0iMuHlKBQzPx...To ensure that ALL of this data gets properly removed, I need to remove this file from all the commits in the repository with git filter-branch. The command git rm --cached git rm docs is not sufficient in this case.(master) $ git filter-branch --tree-filter &#39;rm -f id_rsa&#39; HEADRewrite c69deb9779a30e6335ab1a8ac1a0825cfc9302e4 (6/6)Ref &#39;refs/heads/master&#39; was rewrittenSo far so good, but what about my other branches that have been created?bash(master) $ git checkout new-feature(new-feature) $ lsdrwxrwxr-x  3 sirko sirko  4096 Dec 31 13:20 ./drwx------ 56 sirko sirko 12288 Dec 31 12:37 ../drwxrwxr-x  8 sirko sirko  4096 Dec 31 13:20 .git/-rw-rw-r--  1 sirko sirko  3243 Dec 31 13:20 id_rsa-rw-r--r--  1 sirko sirko   748 Dec 31 12:41 id_rsa.pub-rw-rw-r--  1 sirko sirko    64 Dec 31 13:20 my_document.txtApparently git filter-branch is applying this changes only to the current branch, which is actually not what I want. To make this work, it seems that I have to run git filter-branch in every existing branch, which makes it pretty annoying. After reading more in the (git docs)[https://git-scm.com/docs/git-filter-branch], I found that I need to apply the --all option.(master) $ git filter-branch --tree-filter &#39;rm -f id_rsa&#39; HEAD --allRewrite c69deb9779a30e6335ab1a8ac1a0825cfc9302e4 (7/7)Ref &#39;refs/heads/master&#39; was rewrittenWARNING: Ref &#39;refs/heads/master&#39; is unchangedRef &#39;refs/remotes/origin/master&#39; was rewrittenWARNING: Ref &#39;refs/remotes/origin/master&#39; is unchangedRef &#39;refs/remotes/origin/new-feature&#39; was rewritten(master) $ git checkout new-feature(new-feature) $ lltotal 44drwxrwxr-x  3 sirko sirko  4096 Dec 31 13:41 ./drwx------ 57 sirko sirko 12288 Dec 31 13:41 ../drwxrwxr-x  8 sirko sirko  4096 Dec 31 13:41 .git/-rw-rw-r--  1 sirko sirko   748 Dec 31 13:41 id_rsa.pub-rw-rw-r--  1 sirko sirko    64 Dec 31 13:41 my_document.txtThat seems to be exactly what I want and in the end I just need to git push --all --force my changes. After doing this, all collaborators should dump their local versions and clone a fresh version from the origin.Another alternative to working with git filter-branch is BFG which has some more nifty features.This tool provides some commands to completely remove big files as well as passwords from your Git history. Sadly I could not get it properly working, big files are still persistent as a git object and passwords can not be deleted because they are protected by &#39;HEAD&#39;. I could not really find a solution for these problems. Maybe you are more lucky!The easiest and much simpler solution is to initialize a new git repository, after making sure to have all sensitive information removed. The downside is obviously the loss of the project&#39;s Git history." />
```

Comment:
@etagwerker this looks too long, should be around 155 characters according to https://moz.com/learn/seo/meta-description